File: src/game/botAI.ts
import { GameMap, Player, WeaponPickup, WeaponType } from '../types';
import { WEAPONS, getWeapon } from '../data/weapons';

export function updateBotAI(
  bot: Player,
  players: Player[],
  pickups: WeaponPickup[],
  map: GameMap,
  deltaTime: number,
  onBotShoot: (bot: Player, angle: number) => void,
  onBotThrowGrenade: (bot: Player, angle: number) => void
) {
  if (bot.isDead) return;

  // Find nearest valid enemy target
  let closestTarget: Player | null = null;
  let minDistance = Infinity;

  for (const other of players) {
    if (other.id === bot.id || other.isDead) continue;
    if (bot.team !== 'none' && other.team === bot.team) continue;

    // Ignore enemies hiding deep in bushes unless nearby
    const dist = Math.hypot(other.x - bot.x, other.y - bot.y);
    if (other.inBush && dist > 250) continue;

    if (dist < minDistance) {
      minDistance = dist;
      closestTarget = other;
    }
  }

  // Difficulty parameters
  let accuracyOffset = 0.25;
  let reactionRate = 0.08;
  let grenadeProbability = 0.002;

  switch (bot.botDifficulty) {
    case 'easy':
      accuracyOffset = 0.35;
      reactionRate = 0.04;
      grenadeProbability = 0.001;
      break;
    case 'medium':
      accuracyOffset = 0.22;
      reactionRate = 0.08;
      grenadeProbability = 0.003;
      break;
    case 'hard':
      accuracyOffset = 0.12;
      reactionRate = 0.14;
      grenadeProbability = 0.006;
      break;
    case 'pro':
      accuracyOffset = 0.04;
      reactionRate = 0.25;
      grenadeProbability = 0.01;
      break;
  }

  // Movement inputs initialized
  let moveLeft = false;
  let moveRight = false;
  let boost = false;
  let crouch = false;
  let aimAngle = bot.aimAngle;

  if (closestTarget) {
    const dx = closestTarget.x - bot.x;
    const dy = closestTarget.y - bot.y;
    const distance = Math.hypot(dx, dy);

    // Aim angle calculation with difficulty spread
    const rawAngle = Math.atan2(dy, dx);
    aimAngle += (rawAngle - aimAngle) * reactionRate + (Math.random() - 0.5) * accuracyOffset * 0.1;

    // Tactical positioning
    const desiredDist = getWeapon(bot.primaryWeapon).category === 'melee' ? 30 : 250;

    if (distance > desiredDist + 50) {
      if (dx > 0) moveRight = true;
      else moveLeft = true;
    } else if (distance < desiredDist - 50) {
      if (dx > 0) moveLeft = true;
      else moveRight = true;
    }

    // Vertical Jetpack boosting if target is higher or vertical platform obstacle
    if (dy < -60 || (!bot.isGrounded && Math.random() < 0.3)) {
      boost = true;
    }

    // Crouch dodging if low health or target has high damage weapon
    if (bot.health < 40 && Math.random() < 0.15) {
      crouch = true;
    }

    // Shooting decision
    if (distance < 600 && Math.random() < reactionRate * 2) {
      onBotShoot(bot, aimAngle);
    }

    // Grenade throw decision
    if (distance > 180 && distance < 450 && Math.random() < grenadeProbability) {
      onBotThrowGrenade(bot, aimAngle);
    }
  } else {
    // Idle Wander Patrol
    if (Math.random() < 0.02) {
      bot.facingRight = !bot.facingRight;
    }
    if (bot.facingRight) moveRight = true;
    else moveLeft = true;

    if (Math.random() < 0.01) {
      boost = true;
    }
  }

  // Check weapon pickups nearby to upgrade weapon
  for (const pickup of pickups) {
    if (pickup.respawnTime > 0) continue;
    const distToPickup = Math.hypot(pickup.x - bot.x, pickup.y - bot.y);
    if (distToPickup < 40) {
      // Pick up heavy/better weapon if holding starter pistol
      const currentCategory = getWeapon(bot.primaryWeapon).category;
      const pickupCategory = getWeapon(pickup.weaponType).category;
      if (currentCategory === 'pistol' || pickupCategory === 'heavy' || pickupCategory === 'special') {
        bot.secondaryWeapon = bot.primaryWeapon;
        bot.primaryWeapon = pickup.weaponType;
        pickup.respawnTime = 8.0; // Auto-spawns after 8s
      }
    }
  }

  return {
    moveLeft,
    moveRight,
    boost,
    crouch,
    aimAngle
  };
}

File: server/game/bot_ai.go
package game

import (
	"math"
	"math/rand"
)

func UpdateServerBotAI(bot *Player, players []*Player, pickups []WeaponPickup, gameMap GameMap, deltaTime float64) (ClientInput, bool, bool) {
	if bot.IsDead {
		return ClientInput{}, false, false
	}

	var closestTarget *Player
	minDist := math.Inf(1)

	for _, other := range players {
		if other.ID == bot.ID || other.IsDead {
			continue
		}
		if bot.Team != "none" && other.Team == bot.Team {
			continue
		}

		dist := math.Hypot(other.X-bot.X, other.Y-bot.Y)
		if other.InBush && dist > 250 {
			continue
		}
		if dist < minDist {
			minDist = dist
			closestTarget = other
		}
	}

	accuracyOffset := 0.25
	reactionRate := 0.08
	grenadeProb := 0.003

	switch bot.BotDifficulty {
	case "easy":
		accuracyOffset = 0.35
		reactionRate = 0.04
		grenadeProb = 0.001
	case "medium":
		accuracyOffset = 0.22
		reactionRate = 0.08
		grenadeProb = 0.003
	case "hard":
		accuracyOffset = 0.12
		reactionRate = 0.14
		grenadeProb = 0.006
	case "pro":
		accuracyOffset = 0.04
		reactionRate = 0.25
		grenadeProb = 0.01
	}

	var input ClientInput
	shouldShoot := false
	shouldThrowGrenade := false

	if closestTarget != nil {
		dx := closestTarget.X - bot.X
		dy := closestTarget.Y - bot.Y
		dist := math.Hypot(dx, dy)

		rawAngle := math.Atan2(dy, dx)
		bot.AimAngle += (rawAngle-bot.AimAngle)*reactionRate + (rand.Float64()-0.5)*accuracyOffset*0.1
		input.AimAngle = bot.AimAngle

		desiredDist := 250.0
		if w, ok := Weapons[bot.PrimaryWeapon]; ok && w.Category == "melee" {
			desiredDist = 30.0
		}

		if dist > desiredDist+50 {
			if dx > 0 {
				input.MoveRight = true
			} else {
				input.MoveLeft = true
			}
		} else if dist < desiredDist-50 {
			if dx > 0 {
				input.MoveLeft = true
			} else {
				input.MoveRight = true
			}
		}

		if dy < -60 || (!bot.IsGrounded && rand.Float64() < 0.3) {
			input.Boost = true
		}

		if bot.Health < 40 && rand.Float64() < 0.15 {
			input.Crouch = true
		}

		if dist < 600 && rand.Float64() < reactionRate*2 {
			shouldShoot = true
		}

		if dist > 180 && dist < 450 && rand.Float64() < grenadeProb {
			shouldThrowGrenade = true
		}

		// Check for nearby weapon pickups
		for _, pickup := range pickups {
			if pickup.RespawnTime <= 0 && math.Hypot(pickup.X-bot.X, pickup.Y-bot.Y) < 35.0 {
				if bot.PrimaryWeapon != pickup.WeaponType {
					input.PickUpWeapon = true
					break
				}
			}
		}
	} else {
		// Idle patrol
		if rand.Float64() < 0.02 {
			bot.FacingRight = !bot.FacingRight
		}
		if bot.FacingRight {
			input.MoveRight = true
		} else {
			input.MoveLeft = true
		}
		if rand.Float64() < 0.01 {
			input.Boost = true
		}
		input.AimAngle = bot.AimAngle
	}

	return input, shouldShoot, shouldThrowGrenade
}
